### PR TITLE
add ITT recipe 

### PR DESCRIPTION
**Pull request by jingxu10**
Add tutorial for ITT feature at https://github.com/pytorch/pytorch/pull/63289

### CODE REVIEW

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:54]:
```suggestion
2. Explicit invocation: If customized labeling is needed, users can use APIs mentioned in `PyTorch Docs <https://pytorch.org/docs/stable/profiler.html#intel-instrumentation-and-tracing-technology-apis>`__ explicitly to label a desired range.
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:57]:
```suggestion
To enable this feature, codes which are expected to be labeled should be invoked under a `torch.autograd.profiler.emit_itt()` scope. For example:
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:65]:
```suggestion
To verify the functionality, you need to start an Intel® VTune™ Profiler instance. Please check the `Intel® VTune™ Profiler User Guide <https://www.intel.com/content/www/us/en/develop/documentation/vtune-help/top/launch.html>`__ for steps to launch Intel® VTune™ Profiler.
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:67]:
```suggestion
Once you get the Intel® VTune™ Profiler GUI launched, you should see a user interface as below:
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:73]:
```suggestion
Three sample results are available in the left side navigation bar under `sample (matrix)` project. If you do not want profiling results appear in this default sample project, you can create a new project via the button `New Project...` under the blue `Configure Analysis...` button. To start a new profiling, click the blue `Configure Analysis...` button to initiate configuration of the profiling.
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:79]:
```suggestion
The right side of the windows is split into 3 parts: `WHERE` (top left), `WHAT` (bottom left), and `HOW` (right). With `WHERE`, you can assign a machine where you want to run the profiling on. With `WHAT`, you can set the path of the application that you want to profile. To profile a PyTorch script, it is recommended to wrap all manual steps, including activate a conda environment and setting required environment variable, into a bash script, then profile this bash script. In the screenshot above, we wrapped all steps into the `launch.sh` bash script and profile `bash` with the parameter to be `<path_of_launch.sh>`. In the right side `HOW`, you can choose whatever type that you would like to profile. Details can be found at `Intel® VTune™ Profiler User Guide <https://www.intel.com/content/www/us/en/develop/documentation/vtune-help/top/analyze-performance.html>`__.
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:81]:
```suggestion
With a successful profiling with ITT, you can open `Platform` tab of the profiling result to see labels in the Intel® VTune™ Profiler timeline. All operators starting with `aten::` are operators labeled implicitly by the ITT feature in PyTorch. Labels `iteration_N` are explicitly labeled with specific APIs `torch.profiler.itt.range_push()`, `torch.profiler.itt.range_pop()` or `torch.profiler.itt.range()` scope. Please check the sample code in the next section for details.
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:90]:
```suggestion
The sample code below is the script that was used for profiling in the screenshots above.
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:92]:
```suggestion
The topology is formed by two operators, `Conv2d` and `Linear`. Three iterations of inference were performed. Each iteration was labeled by PyTorch ITT APIs as text string `iteration_N`. Either pair of `torch.profile.itt.range_push` and `torch.profile.itt.range_pop` or `torch.profile.itt.range` scope does the customized labeling feature.
```

File: recipes_source/profile_with_itt.rst
Comment (svekars) [inline recipes_source/profile_with_itt.rst:139]:
```suggestion
   # Retrive the directory path where the path contains both the sample.py and launch.sh so that this script can be invoked from any directory
```

File: recipes_source/recipes_index.rst
Comment (svekars) [inline recipes_source/recipes_index.rst:125]: Also, need to add to the toctree below.

**Review by svekars — COMMENTED** [review]
Some editorial suggestions.

**Comment by jingxu10** [comment]
> Some editorial suggestions.

Done amendments

**Comment by svekars** [comment]
@chauhang can you please assign someone to review this PR? Thanks!

File: recipes_source/profile_with_itt.rst
Comment (malfet) [inline recipes_source/profile_with_itt.rst:65]: English is not my native, but `codes` does not sound like a correct plural for `code`. Perhaps it can be replaced with something like `code to be profiled`
```suggestion
To enable this feature, code block, which is expected to be labeled should be invoked within  `torch.autograd.profiler.emit_itt()` scope. For example:

.. code:: python3

   with torch.autograd.profiler.emit_itt():
     <code-to-be-profiled...>

```

File: recipes_source/profile_with_itt.rst
Comment (malfet) [inline recipes_source/profile_with_itt.rst:146]: Not sure how this section is specific to ITT? User does not need to install (nor use) conda to benefit from ITT feature. Nor should it be bound to python-3.8 , as environment name suggests. And since example does not have any data dependencies, changing working directory to base folder is superfluous as well, isn't it?

But paragraph about ITT filename and folder where it will be generated are probably relevant, isn't it?

**Review by malfet — APPROVED** [review]
LGTM, with two minor suggestions:
 - Get rid of `<codes>`
 - Explain better/get rid of starter script

File: recipes_source/profile_with_itt.rst
Comment (jingxu10) [reply recipes_source/profile_with_itt.rst:146]: To use VTune to profile a python script, it is always recommended to wrap the execution of the python script into a bash script. Because executing a python script requires the configuration of a Python environment, which is not always there by default. Also, launching the script from vtune could be done in a directory that doesn't have the python script, so it is better to switch the active directory into the desired folder inside the bash script execution.

I restated the description to make it more descriptive and related to vtune. Also I changed the specific python version and conda distribution into a generic description. Does the updated content below look good to you?
```
The `launch.sh` bash script, mentioned in the Intel® VTune™ Profiler GUI screenshot, to wrap all manual steps is shown below.

.. code:: bash

   # launch.sh

   #!/bin/bash

   # Retrive the directory path where the path contains both the sample.py and launch.sh so that this bash script can be invoked from any directory
   BASEFOLDER=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
   <Activate a Python environment>
   cd ${BASEFOLDER}
   python sample.py
```

**Comment by jingxu10** [comment]
Hi @malfet , @svekars , all CI had passed. Would you take a final review? Shall I do the merge with pytorchbot or you guys will merge this PR? Thank you.

**Comment by svekars** [comment]
@jingxu10 we are trying to get a review from the partners team. Stay tuned.

File: recipes_source/profile_with_itt.rst
Comment (msaroufim) [inline recipes_source/profile_with_itt.rst:85]: Could you please elaborate a bit more on how to read this picture? What do the big brown boxes mean? Or what does someone learn once they look at this picture?

A big convolution gets broken into multiple smaller ones on different thread?

**Review by msaroufim — APPROVED** [review]

File: recipes_source/profile_with_itt.rst
Comment (HamidShojanazeri) [reply recipes_source/profile_with_itt.rst:75]: are the three dots after "New project..." and " configure Analysis..."  intended?

File: recipes_source/profile_with_itt.rst
Comment (HamidShojanazeri) [inline recipes_source/profile_with_itt.rst:59]: does "this" here refers to the explicit invocation?

File: recipes_source/profile_with_itt.rst
Comment (HamidShojanazeri) [inline recipes_source/profile_with_itt.rst:81]: typos : In the right side `HOW --> on the right side 'How'

Comment (HamidShojanazeri) [reply recipes_source/profile_with_itt.rst:81]: can you please clarify on "  you can choose whatever type that you would like to profile."

**Review by HamidShojanazeri — COMMENTED** [review]
Thanks @jingxu10 this is a great addition to the recipes.

File: recipes_source/profile_with_itt.rst
Comment (jingxu10) [reply recipes_source/profile_with_itt.rst:81]: It's the second way. Users launch VTune first, then launch the application from vtune. `WHAT` in VTune is the place to let vtune know which application to launch and profile.

**Comment by jingxu10** [comment]
> I agree with @malfet's feedback and also the starting section is a bit too long so I'd shorten that substantially and maybe get rid of the second image otherwise looks good. I'll let @svekars handle merging once the feedback is addressed

The second image, with descriptions below it, introduces how to use Intel VTune Profiler to profile a PyTorch script. I'll shorten the starting 2 `What is ...` sections.

File: recipes_source/profile_with_itt.rst
Comment (jingxu10) [reply recipes_source/profile_with_itt.rst:75]: Yes, the three dots are shown in the VTune GUI buttons, so I copied them.

File: recipes_source/profile_with_itt.rst
Comment (jingxu10) [reply recipes_source/profile_with_itt.rst:59]: Yes, this refers to the explicit invocation. I'll make it clear in the tutorial.

File: recipes_source/profile_with_itt.rst
Comment (jingxu10) [reply recipes_source/profile_with_itt.rst:81]: Intel VTune Profiler provides a bunch of profiling types (https://www.intel.com/content/www/us/en/develop/documentation/vtune-help/top/analyze-performance.html) that users can choose from.

**Comment by jingxu10** [comment]
Hi @msaroufim @HamidShojanazeri , would you please review the updated content? Do they align with your expectation?

File: recipes_source/profile_with_itt.rst
Comment (malfet) [inline recipes_source/profile_with_itt.rst:49]: You still have `codes` here

Again, English is not my native, but please compare results of https://www.bing.com/search?q=codes vs https://www.bing.com/search?q=code

Comment (jingxu10) [reply recipes_source/profile_with_itt.rst:49]: Oh, your concern is on the plural. I'm not a native speaker as well. I'll change it to `code` anyway.

File: recipes_source/profile_with_itt.rst
Comment (wenzhe-nrv) [inline recipes_source/profile_with_itt.rst:156]: @jingxu10 is it `Retrieve the directory path` here? Might be a typo.

Comment (jingxu10) [reply recipes_source/profile_with_itt.rst]: Good catch. Corrected.

**Comment by jingxu10** [comment]
If the updates look good to you, please help to merge it.
Thank you.

File: recipes_source/profile_with_itt.rst
Comment (min-jean-cho) [inline recipes_source/profile_with_itt.rst:98]: @jingxu10, utlized -> utilized ?

Comment (jingxu10) [reply recipes_source/profile_with_itt.rst]: done

**Review by svekars — APPROVED** [review]

**Review by HamidShojanazeri — APPROVED** [review]

**Comment by jingxu10** [comment]
Hi, may I know status? Are we waiting for more approvals or it is ok to merge? Please feel free to let me know if there are further comments.

**Comment by svekars** [comment]
@jingxu10 I don't think there are any more comments left and the PR looks good. However, we typically merge tutorials for the new release a couple days before the release. There are no action items from you as of now and I will keep you posted.

**Comment by jingxu10** [comment]
Got it.
Thanks a lot!

**Pull request merged**
